Client/SafeBrowsing/GSafeBrowsingCache.swift

```diff
@@ -0,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+private struct ThreatCache: Hashable {
+    var time: TimeInterval
+    let hash: String
+    let threatType: ThreatType
+}
+
+enum SBCacheHitType {
+    case miss
+    case positive
+    case negative
+}
+
+struct SBCacheResult {
+    let threats: [ThreatType]
+    let cacheResult: SBCacheHitType
+}
+
+class SafeBrowsingCache {
+    private let cacheLock = NSRecursiveLock()
+    private var positiveCache = [String: [ThreatCache]]()
+    private var negativeCache = [String: TimeInterval]()
+    
+    func find(_ hash: String) -> SBCacheResult {
+        cacheLock.lock(); defer { cacheLock.unlock() }
+        
+        if let threats = positiveCache[hash] {
+            var results = [ThreatType]()
+            
+            for threat in threats {
+                if Date().timeIntervalSince1970 > threat.time  {
+                    results.append(threat.threatType)
+                } else {
+                    return SBCacheResult(threats: [], cacheResult: .miss)
+                }
+            }
+            
+            if !results.isEmpty {
+                return SBCacheResult(threats: results, cacheResult: .positive)
+            }
+        }
+        
+        for i in 4..<32 {
```

Comment:
One optimization that could be done here (and is what Firefox does) is to only use 4-byte prefixes whenever we need to call `findFullHashes`. Then we can simply set the negative expiry on the 4-byte prefix since we'll never request a 5-byte prefix from the server.

This trick does work because if you get a match for a 5-byte prefix and then ask for all of the full hashes that start with the first 4 bytes of that prefix, it will include the full hashes that start with the same 5 bytes too.

Reply:
Yes that would work, but I basically did exactly what Google said to support variable length hashes.

For such an optimization I think I will have to verify the results returned with the full length hash which we are doing anyway. I will add this in either this PR or another PR for performance.

Thanks!